src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java

```diff
@@ -0,0 +1,155 @@
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.rules.args.Arg;
+import com.facebook.buck.rules.args.WriteToFileArg;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Ordering;
+import java.io.File;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * Used to expand the macro {@literal $(classpath_abi //some:target)} to the transitive abi's jars
+ * path of that target, expanding all paths to be absolute.
+ */
+public class ClasspathAbiMacroExpander extends BuildTargetMacroExpander<ClasspathAbiMacro> {
+
+  @Override
+  public Class<ClasspathAbiMacro> getInputClass() {
+    return ClasspathAbiMacro.class;
+  }
+
+  @Override
+  protected ClasspathAbiMacro parse(
+      BuildTarget target, CellPathResolver cellNames, ImmutableList<String> input)
+      throws MacroException {
+    return ClasspathAbiMacro.of(parseBuildTarget(target, cellNames, input));
+  }
+
+  private HasClasspathEntries getHasClasspathEntries(BuildRule rule) throws MacroException {
+    if (!(rule instanceof HasClasspathEntries)) {
+      throw new MacroException(
+          String.format(
+              "%s used in classpath_abi macro does not correspond to a rule with a java classpath",
+              rule.getBuildTarget()));
+    }
+    return (HasClasspathEntries) rule;
+  }
+
+  @Override
+  public Arg makeExpandToFileArg(BuildTarget target, String prefix, Arg delegate) {
+    return new ClassPathWriteToFileArg(target, prefix, delegate);
+  }
+
+  private SourcePath getJarPath(BuildRule rule, BuildRuleResolver ruleResolver) {
+    SourcePath jarPath = null;
+
+    if (rule instanceof HasSources && rule instanceof HasJavaAbi) {
+      HasJavaAbi javaAbiRule = (HasJavaAbi) rule;
+      Optional<BuildTarget> optionalBuildTarget = javaAbiRule.getAbiJar();
+      if (optionalBuildTarget.isPresent()) {
+        jarPath = ruleResolver.requireRule(optionalBuildTarget.get()).getSourcePathToOutput();
+      }
+    }
+
+    if (jarPath == null) {
+      jarPath = rule.getSourcePathToOutput();
+    }
+
+    return jarPath;
+  }
+
+  @Override
+  protected Arg expand(SourcePathResolver resolver, ClasspathAbiMacro macro, BuildRule rule)
+      throws MacroException {
+    throw new MacroException(
+        "expand(BuildRuleResolver ruleResolver, ClasspathAbiMacro input) should be called instead");
+  }
+
+  @Override
+  public Arg expandFrom(
+      BuildTarget target,
+      CellPathResolver cellNames,
+      BuildRuleResolver resolver,
+      ClasspathAbiMacro input)
+      throws MacroException {
+
+    BuildRule inputRule = resolve(resolver, input);
+    return expand(resolver, inputRule);
+  }
+
+  protected Arg expand(BuildRuleResolver ruleResolver, BuildRule inputRule) throws MacroException {
+
+    ImmutableList<SourcePath> jarPaths =
+        getHasClasspathEntries(inputRule)
+            .getTransitiveClasspathDeps()
```

Comment:
`getTransitiveClasspathDeps` returns both `#aar_prebuilt_jar,class-abi` and `#class-abi` flavours. I verified that both the output jar's are same, just on different paths.

//cache:com.support-compat-27.0.2.aar#aar_prebuilt_jar,class-abi
buck-out/gen/cache/com.support-27.0.2.aar#aar_prebuilt_jar,class-abi/com.support-27.0.2.aar-abi.jar

//cache:com.support-compat-27.0.2.aar#class-abi
buck-out/gen/cache/com.support-27.0.2.aar#class-abi/com.support-27.0.2.aar-abi.jar

How can I filter and keep only the `#class-abi` deps? Also is there a scenario where the output of the two flavours will differ?

Reply:
added a filter to only include rules which have non-null `sourcePathToOutput`, same behaviour is in `ClasspathMacro`.